pandas/_libs/tslibs/tzconversion.pyx

```diff
@@ -566,3 +529,71 @@ cdef int64_t _tz_convert_tzlocal_utc(int64_t val, tzinfo tz, bint to_utc=True,
         return val - delta
     else:
         return val + delta
+
+
+@cython.freelist(16)
+cdef class Localizer:
+    # cdef:
+    #    tzinfo tz
+    #    bint use_utc
+    #    bint use_fixed
+    #    bint use_tzlocal
+    #    bint use_pytz
+    #    bint use_dst
+    #    ndarray trans
+    #    int64_t[:] deltas
+    #    int64_t delta
+    #    str typ
+
+    def __cinit__(self, tzinfo tz):
+        self.tz = tz
+        if is_utc(tz) or tz is None:
+            self.use_utc = True
+        elif is_tzlocal(tz):
+            self.use_tzlocal = True
+        else:
+            trans, deltas, typ = get_dst_info(tz)
+            self.trans = trans
+            self.deltas = deltas
+            self.typ = typ
+
+            if typ not in ["pytz", "dateutil"]:
+                # static/fixed; in this case we know that len(delta) == 1
+                self.use_fixed = True
+                self.delta = deltas[0]
+            else:
+                self.use_dst = True
+                if typ == "pytz":
+                    self.use_pytz = True
+
+    cdef intp_t* prepare1(self, int64_t utc_val):
+        cdef:
+            intp_t loc
+
+        if self.use_dst:
+            loc = self.trans.searchsorted(utc_val, side="right") - 1
+            return &loc
```

Comment:
I think this is the problematic line. I’m a traditional C function, loc would go out of scope when the function ends. So returning the address of loc gives a pointer to a location that may very well be freed when the function returns.